WordPress/src/main/java/org/wordpress/android/util/EncryptedLogging.kt

```diff
@@ -0,0 +1,92 @@
+import org.wordpress.android.modules.BG_THREAD
+import org.wordpress.android.util.AppLog.T
+import org.wordpress.android.util.analytics.AnalyticsTrackerWrapper
+import java.io.File
+import java.util.UUID
+import javax.inject.Inject
+import javax.inject.Named
+import javax.inject.Singleton
+
+@Singleton
+class EncryptedLogging @Inject constructor(
+    private val dispatcher: Dispatcher,
+    private val encryptedLogStore: EncryptedLogStore,
+    private val networkUtilsWrapper: NetworkUtilsWrapper,
+    private val analyticsTrackerWrapper: AnalyticsTrackerWrapper,
+    @Named(BG_THREAD) private val bgDispatcher: CoroutineDispatcher
+) {
+    private val coroutineScope = CoroutineScope(bgDispatcher)
+
+    init {
+        dispatcher.register(this)
+    }
+
+    fun start() {
+        dispatcher.dispatch(EncryptedLogActionBuilder.newResetUploadStatesAction())
+        if (networkUtilsWrapper.isNetworkAvailable()) {
+            coroutineScope.launch {
+                encryptedLogStore.uploadQueuedEncryptedLogs()
+            }
+        }
+    }
+
+    /**
+     * Dispatches a FluxC action that will queue the given log to be uploaded as soon as possible.
+     *
+     * @param logFile Log file to be uploaded
+     * @param shouldStartUploadImmediately This parameter will decide whether we should try to upload the log file
+     * immediately. After a crash, we are unlikely to have enough time to complete the upload, so we can use this
+     * parameter to avoid the unnecessary upload failure.
+     */
+    fun encryptAndUploadLogFile(logFile: File, shouldStartUploadImmediately: Boolean): String? {
+        if (logFile.exists()) {
+            val uuid = UUID.randomUUID().toString()
+            val payload = UploadEncryptedLogPayload(
+                    uuid = uuid,
+                    file = logFile,
+                    // If the connection is not available, we shouldn't try to upload immediately
+                    shouldStartUploadImmediately = shouldStartUploadImmediately &&
+                            networkUtilsWrapper.isNetworkAvailable()
+            )
+            dispatcher.dispatch(EncryptedLogActionBuilder.newUploadLogAction(payload))
+            return uuid
+        }
+        return null
+    }
+
+    @Suppress("unused")
+    @Subscribe(threadMode = ASYNC)
+    fun onEncryptedLogUploaded(event: OnEncryptedLogUploaded) {
+        when (event) {
+            is EncryptedLogUploadedSuccessfully -> {
+                AppLog.i(T.MAIN, "Encrypted log with uuid: ${event.uuid} uploaded successfully!")
+                analyticsTrackerWrapper.track(Stat.ENCRYPTED_LOGGING_UPLOAD_SUCCESSFUL)
+            }
+            is EncryptedLogFailedToUpload -> {
+                AppLog.e(T.MAIN, "Encrypted log with uuid: ${event.uuid} failed to upload with error: ${event.error}")
+                // Only track final errors
+                if (event.willRetry) {
```

Comment:
Just to double check -> shouldn't this be `!event.willRetry`?

If it's correct the way it is now, I feel like the comment might be a bit confusing.

Reply:
Thanks for catching that @malinajirka! Fixed in 574bbaea5e7399830c3b1600dc5e6822a9d7118a.